src/slowlog.c

```diff
@@ -0,0 +1,160 @@
+#include <assert.h>
+#include <string.h>
+#include "alloc.h"
+#include "corvus.h"
+#include "slowlog.h"
+#include "logging.h"
+
+// use 32bit to guarantee slowlog id will not overflow in slowlog_entry.id
+static uint32_t slowlog_id = 0;
+static inline size_t min(size_t a, size_t b) { return a < b ? a : b; }
+
+
+int slowlog_init(struct slowlog_queue *slowlog)
+{
+    size_t queue_len = 1 + (config.slowlog_max_len - 1) / config.thread;  // round up
+    slowlog->capacity = queue_len;
+    slowlog->entries = cv_calloc(queue_len, sizeof(struct slowlog_entry));
+    slowlog->entry_locks = cv_malloc(queue_len * sizeof(pthread_mutex_t));
+    slowlog->curr = 0;
+    int err;
+    for (size_t i = 0; i != queue_len; i++) {
+        if ((err = pthread_mutex_init(slowlog->entry_locks + i, NULL)) != 0) {
+            LOG(ERROR, "pthread_mutex_init: %s", strerror(err));
+            return CORVUS_ERR;
+        }
+    }
+    return CORVUS_OK;
+}
+
+// Should be called only after all worker threads have stopped
+void slowlog_free(struct slowlog_queue *slowlog)
+{
+    for (size_t i = 0; i != slowlog->capacity; i++) {
+        // no other worker threads will read entry queue any more
+        if (slowlog->entries[i] != NULL) {
+            slowlog_dec_ref(slowlog->entries[i]);
+        }
+        pthread_mutex_destroy(slowlog->entry_locks + i);
+    }
+    cv_free(slowlog->entries);
+    cv_free(slowlog->entry_locks);
+}
+
+struct slowlog_entry *slowlog_create_entry(struct command *cmd, int64_t latency)
+{
+    struct slowlog_entry *entry = cv_calloc(1, sizeof(struct slowlog_entry));
+    entry->id = ATOMIC_INC(slowlog_id, 1);
+    entry->log_time = time(NULL);  // redis also uses time(NULL);
```

Comment:
clock_gettime(CLOCK_MONOTONIC, ...) return a time from an unspecified starting point, it's just used to calculate time interval and not suitable here.